### PR TITLE
auth gmysql: modernise implementation of gmysql-ssl flag

### LINKED ISSUE #12528: Auth: Generic MySQL backend with incorrectly configured ssl=yes hangs the Auth process
- Program: Authoritative
 - Issue type: Bug report

### Short description
If you configure Auth with  `gmysql-ssl=yes` but your MySQL database does not support SSL, I would expect errors to be logged by Auth and for Auth to either run without the MySQL backend or exit. Currently, it causes Auth to appear to be running and no errors are logged. The logging includes these lines:

```
Feb 10 09:39:26 UDP server bound to 0.0.0.0:5353
Feb 10 09:39:26 TCP server bound to 0.0.0.0:5353
Feb 10 09:39:26 [webserver] Listening for HTTP requests on [::]:8083
```

 However, Auth does not respond to:
- Queries over UDP
- Queries over TCP
- Request made to the webserver

It appears the Auth process is stuck and even after 10+ minutes the Auth logging remains silent.

### Environment

 - Operating system: N/A
 - Software version: 4.7.3
 - Software source: powerdns/pdns-auth-47:4.7.3 at dockerhub

### Steps to reproduce

1. Run MySQL database with no ssl support
2. Run Auth with `gmysql-ssl=yes`
3. Try to access Do53 TCP/UDP and webserver

### Other information
Changing the config to `gmysql-ssl=no` (the correct configuration for this test setup) and restarting Auth leads to a properly functioning Auth. Modifying parameters such as `zone-cache-refresh-interval` & `loglevel` did not lead to a difference in behaviour and/or relevant information being logged.

### LINKED ISSUE #12978: Authoritative server 4.7.4 on RHEL8 can't use SSL with mariadb
- Program: Authoritative version 4.7.4, using mysql-backend (mariadb 10.11.4) on RHEL8 
 - Issue type: Bug report

### Short description

Trying to start an authoritative 4.7.4 server on RHEL8, with gmysql-backend (mariadb 10.11.4, also on RHEL8).
However, every time I try to start the service, I recieve the following issue:
```
gmysql Connection failed: Unable to connect to database: ERROR 1043 (08S01): Bad handshake
Caught an exception instantiating a backend: Unable to launch gmysql connection: Unable to connect to database: ERROR 1043 (08S01): Bad handshake
```

### Environment

 - Operating system: RHEL8
 - Software version: PDNS authoritative 4.7.4, mariadb 10.11.4
 - Software source: PDNS repo`https://repo.powerdns.com/el/$basearch/$releasever/auth-47/debug`
 Mariadb repo: ` https://mirror.yer.az/mariadb/yum/10.11/rhel/$releasever/$basearch`
 
 PDNS.conf file:
 ```
 launch=gmysql

gmysql-host=192.168.121.162
gmysql-user=pdnsuser
gmysql-dbname=pdns
gmysql-password=pdnsuser123
gmysql-ssl=yes
api=yes
api-key=76c76bcc2e870735c
webserver-address=0.0.0.0
webserver-allow-from=192.168.0.0/16,127.0.0.1,10.0.0.0/8
security-poll-suffix=
loglevel=5
log-dns-queries=yes

#resolver=[::1]:53
expand-alias=yes
default-soa-content=a.misconfigured.dns.server.invalid hostmaster.@ 0 10800 3600 604800 300
 ```

mariadb has a database called `pdns` with all necessary tables (created via the installation guide).
it is important to note, I can use `mysql` to access the mariadb server from the PDNS auth machine just fine:
```
[vagrant@pdns ~]$ 
[vagrant@pdns ~]$ mysql -h 192.168.121.162 -updnsuser -p
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 21
Server version: 10.11.4-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> 
MariaDB [(none)]>
```
### Steps to reproduce

1. create 2 RHEL8 nodes, one with mariadb 10.11.4, one with PDNS authoritative server 4.7.4
2. Try to use the mariadb as a backend for the PDNS server (With SSL!) and receive the same error when trying to run the pdns server.

### Expected behaviour
PDNS server running without any issue.

### Actual behaviour
receiving this `SSL handshake` error upon startup, and failing.

### Other information
I've already discussed with the IRC support, but I could not find any workaround:
1. using  `update-crypto-policies` does not solve the issue.
2. Trying to install a newer version of `mariadb-connector-c` is impossible, since the latest version which is support in EL8 is `mariadb-connector-c.x86_64 3.1.11-2.el8_3`.

### PR DESCRIPTION
**Pull request by Habbie**
### Short description
Fixes #12978, fixes #12528

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by Habbie** [comment]
`smysql.cc:511:28: error: ‘MYSQL_OPT_SSL_ENFORCE’ was not declared in this scope; did you mean ‘MYSQL_OPT_SSL_MODE’?`

File: docs/backends/generic-mysql.rst [hunk 1]
```diff
@@ -136,7 +136,11 @@ Use the InnoDB READ-COMMITTED transaction isolation level. Default: yes.
 ``gmysql-ssl``
 ^^^^^^^^^^^^^^^^^^
 
-Send the CLIENT_SSL capability flag to the server. SSL support is announced by the server via CLIENT_SSL and is enabled if the client returns the same capability. Default: no.
+.. versionchanged:: 4.9.0
+
+Before 4.9.0: Send the CLIENT_SSL capability flag to the server. SSL support is announced by the server via CLIENT_SSL and is enabled if the client returns the same capability. Default: no.
```
Comment (zeha) [inline docs/backends/generic-mysql.rst:141]: `Default: no` wants to be in its own line, after all version things?

Comment (zeha) [reply docs/backends/generic-mysql.rst:141]: (I feel like a statement ala "The exact meaning of these flags are MySQL-client and -version dependent" might also be a good C-Y-A line...)

**Comment by securitym0nkey** [comment]
> New plan: document how to configure client side SSL in `my.cnf` via `gmysql-group` and either get rid of `-ssl`, make it a no-op, or leave it as is for the unlikely event somebody needs the old behaviour.

I found your pull request while looking to enforce certificate validation. Reading thru the mysql docs i found out that these days you should use **ssl-mode**. Most part work fine with the my.cnf file. Like setting **ssl-ca** or **ssl-capath**. But unfortunately it seems not possible to set the ssl-mode via the my.cnf. The old "ssl-verify-server-cert" is also ignored. So i think there a need to set the ssl-mode from the qmysql site.

**Comment by Habbie** [comment]
> So i think there a need to set the ssl-mode from the qmysql site.

It turns out there is no reliable way to do that from our side, between old libmysqlclient, new libmysqlclient, and mariadb. If the client library cannot do it from the config, I'm afraid the client library will need to be fixed.

**Comment by Habbie** [comment]
This is ready for review now.

**Review by rgacogne — APPROVED** [review]

**Pull request merged**